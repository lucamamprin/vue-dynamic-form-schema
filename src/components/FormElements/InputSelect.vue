<template>
  <div
    class="input select floating-label floating-label-floated"
    :class="{'input-transparent': darkMode,}"
  >
    <select
      :id="id"
      :class="{
        'border-danger': hasError,
        'is-selected': value,
      }"
      :aria-describedby="describedBy"
      :name="name"
      :disabled="disabled"
      :aria-invalid="invalid.toString()"
      :aria-required="required.toString()"
      :aria-labelledby="`label-${name}`"
      @change="$emit('input', {
        label: $event.target.options[$event.target.options.selectedIndex].innerText,
        value: $event.target.value
      })"
    >
      <option
        value
        :selected="!value"
      >
        {{ firstValue }}
      </option>
      <option
        v-for="opt in options"
        :key="opt.value"
        :value="opt.value"
        :selected="opt.value === value"
      >
        {{ opt.label }}
      </option>
    </select>

    <label
      v-if="label"
      :id="`label-${name}`"
      :for="id"
      v-html="label"
    />

    <div v-if="hasError && errorMessage">
      <error-message
        :id="describedBy"
        :error-message="errorMessage"
      />
    </div>
  </div>
</template>

<script>
import props from "../../mixins/props"
import options from "../../mixins/options"
const ErrorMessage = () => import("./Atoms/ErrorMessage")

export default {
  components: {
    ErrorMessage,
  },
  mixins: [
    props,
    options,
  ],
  props: {
    firstValue: {
      type: String,
      default: "",
    },
    value: {
      type: [
        String, Number, Object,
      ],
      default: null,
    },
  },
  data () {
    return {
      val: null,
    }
  },
}
</script>
