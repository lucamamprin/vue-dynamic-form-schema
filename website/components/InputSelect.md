# InputSelect

## Props

<!-- @vuese:InputSelect:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|id|-|`String`|`true`|-|
|describedBy|-|`String`|`true`|-|
|name|-|`String`|`true`|-|
|label|-|`String`|`true`|-|
|firstValue|-|`String`|`false`|-|
|value|-|`String` /  `Number` /  `Object`|`false`|null|
|disabled|-|`Boolean`|`false`|false|
|options|-|`Array`|`true`|-|
|hasError|-|`Boolean`|`false`|false|
|errorMessage|-|`String`|`false`|-|
|darkMode|-|`Boolean`|`false`|false|
|invalid|-|`Boolean`|`false`|-|
|required|-|`Boolean`|`false`|false|
|resetMarginTop|-|`Boolean`|`false`|false|

<!-- @vuese:InputSelect:props:end -->


