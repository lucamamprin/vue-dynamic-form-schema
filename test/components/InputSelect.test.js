import Vue from "vue"
import { shallowMount } from "@vue/test-utils"
import InputSelect from "../../src/components/FormElements/InputSelect.vue"

const props = {
  id: "favoriteAnimal",
  describedBy: "favoriteAnimal-describe",
  name: "animal",
  label: "Favorite animal",
  required: true,
  options: [

  ],
}

const options = [
  {
    "label": "Dog",
    "value": "dog",
  },
  {
    "label": "Cat",
    "value": "cat",
  },
]

const propsValues = {
  ...props,
  options,
}

let wrapper

describe("InputSelect", () => {
  beforeEach(() => {
    wrapper = shallowMount(InputSelect, {
      propsData: props,
    })
  })
  
  test("mounts correctly", () => {
    expect(wrapper.vm).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })

  test("gets array of options", async () => {
    await wrapper.setProps({
      ...propsValues,
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test("should be disabled", async () => {
    await wrapper.setProps({
      disabled: true,
    })

    expect(wrapper.find("select").element.disabled).toBeTruthy()
  })

  test("sets new value", async () => {
    await wrapper.setProps({
      ...propsValues,
    })

    const options = wrapper.find("select").findAll("option")

    await options.at(1).setSelected()

    expect(wrapper.find("option:checked").element.value).toBe("dog")
    expect(wrapper.find("select").element.value).toBe("dog")
  })

  test("shows error message", async () => {
    await wrapper.setProps({
      invalid: true,
      hasError: true,
      errorMessage: "You need to select an animal.",
    })

    await Vue.nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  test("change event emitted", async () => {
    wrapper.vm.$emit("input", {
      label: "Seal",
      value: "seal",
    })

    await Vue.nextTick()

    const event = wrapper.emitted("input")
    const value = event[0][0]

    expect(wrapper.emitted("input")).toBeTruthy()

    expect(typeof value).toBe("object")
  })
})
