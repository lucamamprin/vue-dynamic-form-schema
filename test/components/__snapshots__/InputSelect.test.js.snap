// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputSelect gets array of options 1`] = `
<div class="input select floating-label floating-label-floated"><select id="favoriteAnimal" aria-describedby="favoriteAnimal-describe" name="animal" aria-invalid="false" aria-required="true" aria-labelledby="label-animal" class="">
    <option value="">

    </option>
    <option value="dog">
      Dog
    </option>
    <option value="cat">
      Cat
    </option>
  </select> <label id="label-animal" for="favoriteAnimal">Favorite animal</label>
  <!---->
</div>
`;

exports[`InputSelect mounts correctly 1`] = `
<div class="input select floating-label floating-label-floated"><select id="favoriteAnimal" aria-describedby="favoriteAnimal-describe" name="animal" aria-invalid="false" aria-required="true" aria-labelledby="label-animal" class="">
    <option value="">

    </option>
  </select> <label id="label-animal" for="favoriteAnimal">Favorite animal</label>
  <!---->
</div>
`;
